larq_compute_engine/core/packbits_aarch64.h

```diff
@@ -10,53 +10,299 @@
+      // cetera, but also that the sign bits are correctly ordered within each
+      // register.
+      //     The first computation step is a series of UZP2 instructions on
+      // pairs of registers that hold adjacent sign-bits. Every 'odd' byte is
+      // taken from the two registers and concatenated together, preserving
+      // order, in the destination register.
+      //     After this block, the register sequence v16, v18, ..., v30 will
+      // contain the sign bits (in correct order). Because we take every 'odd'
+      // byte from each pair, we are effectively concentrating the sign bits, so
+      // there will be 16 in each register, in the high-bit of each byte.
+      "uzp2 v16.16b, v1.16b, v3.16b\n"
+      "ld2 {v0.8h, v1.8h}, [%[in_ptr]], #32\n"
+      "uzp2 v18.16b, v5.16b, v7.16b\n"
+      "uzp2 v20.16b, v9.16b, v11.16b\n"
+      "ld2 {v2.8h, v3.8h}, [%[in_ptr]], #32\n"
+      "uzp2 v22.16b, v13.16b, v15.16b\n"
+      "uzp2 v24.16b, v17.16b, v19.16b\n"
+      "ld2 {v4.8h, v5.8h}, [%[in_ptr]], #32\n"
+      "uzp2 v26.16b, v21.16b, v23.16b\n"
+      "uzp2 v28.16b, v25.16b, v27.16b\n"
+      "ld2 {v6.8h, v7.8h}, [%[in_ptr]], #32\n"
+      "uzp2 v30.16b, v29.16b, v31.16b\n"
+
+      // The next step is to perform two unzips of each adjacent pair of
+      // sign-bit registers. The first (UZP2) takes 'odd' bytes as before, while
+      // the second (UZP1) takes 'even' bytes.
+      //
+      // To illustrate:
+      //
+      // v16 and v18 contain the labelled bytes as follows (where each byte
+      // contains a single sign-bit):
+      //                        v16          v18
+      //                   [0 1 ... 15] [16 17 ... 31]
+      //
+      // Afterwards, v0 and v17 will contain:
+      //                        v0           v17
+      //                   [1 3 ... 31] [0 2 ... 30]
+      "uzp2 v0.16b, v16.16b, v18.16b\n"
+      "ld2 {v8.8h, v9.8h}, [%[in_ptr]], #32\n"
+      "uzp1 v17.16b, v16.16b, v18.16b\n"
+      "uzp2 v2.16b, v20.16b, v22.16b\n"
+      "ld2 {v10.8h, v11.8h}, [%[in_ptr]], #32\n"
+      "uzp1 v19.16b, v20.16b, v22.16b\n"
+      "uzp2 v4.16b, v24.16b, v26.16b\n"
+      "ld2 {v12.8h, v13.8h}, [%[in_ptr]], #32\n"
+      "uzp1 v21.16b, v24.16b, v26.16b\n"
+      "uzp2 v6.16b, v28.16b, v30.16b\n"
+      "ld2 {v14.8h, v15.8h}, [%[in_ptr]], #32\n"
+      "uzp1 v23.16b, v28.16b, v30.16b\n"
+
+      "sub %[n], %[n], #2\n"
+
+      // Now, we can combine the two unzip results for each register pair, with
+      // 'shift right and insert' SRI instructions.
+      //
+      // Continuing with the example from above, after the first SRI instruction
+      // below, v0 will contain:
+      //                                v0
+      //                     [0/1  2/3  4/5 ... 30/31]
+      //
+      // Where `0/1` is a byte that has the high-bit equal to the high-bit of
+      // byte 0, and the second highest bit equal to the high-bit of byte 1.
```

Comment:
I think the highest bit would be the high-bit of byte 1, and the second highest bit would be the high-bit of byte 0:

```
bit pos:  7  6 5 4 3 2 1 0
value:   s1 s0 _ _ _ _ _ _
```
where the `_` are irrelevant dummy bits.
They'll keep getting right-shifted (i.e. in the direction of bit-position 0) until `s0` is at the least-significant-bit.

Reply:
Ah yes, you're right, thanks for spotting that. Will update :+1: